driver-core/src/main/com/mongodb/connection/LoggingCommandEventSender.java

```diff
@@ -0,0 +1,172 @@
+        if (loggingRequired()) {
+            logger.debug(
+                    format("Sending command '%s ...' with request id %d to database %s on connection [%s] to server %s",
+                            getTruncatedJsonCommand(), message.getId(),
+                            message.getNamespace().getDatabaseName(), description.getConnectionId(), description.getServerAddress()));
+        }
+
+        if (eventRequired()) {
+            BsonDocument commandDocumentForEvent = (securitySensitiveCommands.contains(commandName))
+                    ? new BsonDocument() : commandDocument;
+
+            sendCommandStartedEvent(message, message.getNamespace().getDatabaseName(),
+                    commandName, commandDocumentForEvent, description, commandListener);
+        }
+        // the buffer underlying the command document may be released after the started event, so set to null to ensure it's not used
+        // when sending the failed or succeeded event
+        commandDocument = null;
+    }
+
+    private String getTruncatedJsonCommand() {
+        StringWriter writer = new StringWriter();
+
+        BsonReader bsonReader = commandDocument.asBsonReader();
+        try {
+            JsonWriter jsonWriter = new JsonWriter(writer,
+                    JsonWriterSettings.builder().outputMode(JsonMode.RELAXED).maxLength(MAX_COMMAND_DOCUMENT_LENGTH_TO_LOG).build());
+
+            jsonWriter.pipe(bsonReader);
+
+            return writer.toString();
+        } finally {
+            bsonReader.close();
+        }
+    }
+
+    @Override
+    public void sendFailedEvent(final Throwable t) {
+        Throwable commandEventException = t;
+        if (t instanceof MongoCommandException && (securitySensitiveCommands.contains(commandName))) {
+            commandEventException = new MongoCommandException(new BsonDocument(), description.getServerAddress());
+        }
+        long elapsedTimeNanos = System.nanoTime() - startTimeNanos;
+
+        if (loggingRequired()) {
+            logger.debug(
+                    format("Execution of command with request id %d failed to complete successfully in %.2f ms on connection [%s] "
+                                    + "to server %s",
+                            message.getId(), nanosToMillis(elapsedTimeNanos), description.getConnectionId(),
+                            description.getServerAddress()),
+                    commandEventException);
+        }
+
+        if (eventRequired()) {
+            sendCommandFailedEvent(message, commandName, description, elapsedTimeNanos, commandEventException, commandListener);
+        }
+    }
+
+    @Override
+    public void sendSucceededEvent(final ResponseBuffers responseBuffers) {
+        long elapsedTimeNanos = System.nanoTime() - startTimeNanos;
+
+        if (loggingRequired()) {
+            logger.debug(
+                    format("Execution of command with request id %d completed successfully in %.2f ms on connection [%s] to server %s",
```

Comment:
Logging the message id and elapsed time means that log analyzers (human or machine) don't have to rely on timestamps and thread identifiers (which don't help for async) in log statements to determine the elapsed time of a command.

Reply:
Thats great :+1: